### PR TITLE
tr: implement complement separately from delete or squeeze (#2147)

### LINKED ISSUE #2147: tr: support -c (complement) when translating
Add support translating complement of a character set.

What happens now:
```
$ echo -n ab | target/debug/coreutils tr -c a X
tr: error: -c is only supported with -d or -s
```

What I expected:
```
$ echo -n ab | tr -c a X
aX
```

### PR DESCRIPTION
**Pull request by jhscheer**
fixes #2147

### CODE REVIEW

**Comment by cbjadwani** [comment]
When using ranges in `set2` this will give different output than POSIX and GNU `tr`:

```
$ echo '0x1y2z3' | tr -c '\0-@' '*-~'
0a1b2c3

$ echo '0x1y2z3' | ./target/debug/tr -c '\0-@' '*-~'
0j1j2j3
```

**Comment by jhscheer** [comment]
Thanks for pointing that out.
I pushed a partial fix for this (see new tests).

Now it works if called like this `tr -c '0-@' '*-~'`. 
However, if called like in your example, with a `\0` in set1, then it still doesn't work.

I haven't figured out why GNU `tr` behaves this way, which makes it difficult to create a fix.

**Comment by sylvestre** [comment]
The CI isn't super happy
```

error[E0061]: this function takes 4 arguments but 3 arguments were supplied
   --> src/uu/tr/src/tr.rs:185:24
    |
185 |             translate: TranslateOperation::new(set1, set2, truncate),
    |                        ^^^^^^^^^^^^^^^^^^^^^^^ ----  ----  -------- supplied 3 arguments
    |                        |
    |                        expected 4 arguments
    |
note: associated function defined here
   --> src/uu/tr/src/tr.rs:131:8
    |
131 |     fn new(
    |        ^^^
132 |         set1: ExpandSet,
    |         ---------------
133 |         set2: &mut ExpandSet,
    |         --------------------
134 |         truncate: bool,
    |         --------------
135 |         complement: bool,
    |         ----------------

error: aborting due to previous error
```

**Comment by sylvestre** [comment]
As ride along, could you please fix
```
src/uu/tr/src/tr.rs#L221

WARNING: `cargo clippy`: you checked before that `unwrap()` cannot fail, instead of checking and unwrapping, it's better to use `if let` or `match`
``` 
?

File: tests/by-util/test_tr.rs [hunk 1]
```diff
@@ -59,6 +59,53 @@ fn test_delete_complement_2() {
...
+fn test_complement4() {
+    // $ echo -n '0x1y2z3' | tr -c '0-@' '*-~'
+    // 0~1~2~3
+    new_ucmd!()
+        .args(&["-c", "0-@", "*-~"])
+        .pipe_in("0x1y2z3")
+        .run()
+        .stdout_is("0~1~2~3");
+
+    // TODO: fix this
```
Comment (sylvestre) [inline tests/by-util/test_tr.rs:99]: in a separate PR is fine but could you please move that as a test and marked it as #ignore ? thanks

Comment (jhscheer) [reply tests/by-util/test_tr.rs:99]: done

**Comment by hbina** [comment]
I am pretty sure this is the correct behavior here
https://github.com/uutils/coreutils/blob/e86c989184cbb6777a9f804fa5fe19f6ce680ba2/tests/by-util/test_tr.rs#L101

I tested in Ubuntu and BSD, and they both does the same thing.

The thing about escape code `\` is that if its not an octal, then its just going to become the next character. So `\h` expands to just `h`.  I think there's no ambiguity here because this logic should take precedence over expansion of `0-@`.

>While a backslash followed by a character not listed above is interpreted as that character

See https://www.gnu.org/software/coreutils/manual/html_node/Character-sets.html#Character-sets

**Comment by jhscheer** [comment]
@hbina I'm very confused. 
There is no doubt that the line you quoted is the correct behaviour. That's precisely why it is there, i.e. to document that in this particular case `uu_id` needs to be fixed because it still does not mimic the behaviour of GNU's tr.

**Comment by hbina** [comment]
Oh, sorry 😅 . Singling out "GNU tr" makes me think that GNU tr in particular is misbehaving. My bad.

**Comment by hbina** [comment]
Seems like this "ignoring the backslash" only works for ranges
```bash
hbina@DESKTOP-2D0I5MJ:~/git/uutils$ echo -n "ABCDEFGH" | tr "\A-^" "a-~"
abcdefghhbina@DESKTOP-2D0I5MJ:~/git/echo "lower" | tr "\[:lower:]" "\[:upper:]"
upper
````

**Pull request merged**